### PR TITLE
Feat upgrade php version (8.0)

### PR DESCRIPTION
**Pull request by eldadfux**
## What does this PR do?

Upgrading PHP version from 7.4 to 8.0. Currently using RC2 version.

## Test Plan

Using current tests.

## Related PRs and Issues

None.

### Have you read the [Contributing Guidelines on issues]

Yes

### CODE REVIEW

File: src/Appwrite/Extend/PDOStatement.php
Comment (SoftCreatR) [inline src/Appwrite/Extend/PDOStatement.php:103]: Could it be, that you forgot to add the parameters to `fetchAll()`? You may also restore `$ctor_args` and pass them, too.

**Review by SoftCreatR — COMMENTED** [review]
Not really a problem, but you are attempting to install `docker-cli` twice (line 136 & 141 in Dockerfile).

File: src/Appwrite/Extend/PDOStatement.php
Comment (eldadfux) [reply src/Appwrite/Extend/PDOStatement.php:103]: I'll add both return types and doc blocks 👍🏻

Seems like in PHP8 this method signature was changed to: `(int $fetch_style = PDO::FETCH_BOTH, mixed ...$fetch_args)` I only changed it because it was throwing errors.

**Comment by eldadfux** [comment]
> Not really a problem, but you are attempting to install `docker-cli` twice (line 136 & 141 in Dockerfile).

Fixed 👌🏻

File: src/Appwrite/Extend/PDOStatement.php
Comment (SoftCreatR) [reply src/Appwrite/Extend/PDOStatement.php:103]: This should work fine:

```php
public function fetchAll(int $fetch_style = PDO::FETCH_BOTH, mixed ...$fetch_args, array $ctor_args = []): array
{
    return $this->PDOStatement->fetchAll($fetch_style, $fetch_args, $ctor_args);
}
```

Comment (eldadfux) [reply src/Appwrite/Extend/PDOStatement.php:103]: Yes, but this function can also return false. 🤔 I'm also not sure the $ctor_args is still supported as in PHP8 $fetch_args is now using the splat operator.

Comment (SoftCreatR) [reply src/Appwrite/Extend/PDOStatement.php:103]: > Yes, but this function can also return false. 🤔 I'm also not sure the $ctor_args is still supported as in PHP8 $fetch_args is now using the splat operator.

Well, I haven't found any information regarding a signature change of this method. And yes, you are right, that it can also return false (seems to be a fault in the official documentation). In this case, this should fit:

```php
public function fetchAll(int $fetch_style = PDO::FETCH_BOTH, mixed ...$fetch_args): mixed
{
    return $this->PDOStatement->fetchAll($fetch_style, $fetch_args);
}
```

Comment (SoftCreatR) [reply src/Appwrite/Extend/PDOStatement.php:111]: With proper type hinting, the doc block becomes obsolete :P

File: Dockerfile
Comment (SoftCreatR) [inline Dockerfile:66]: The official YAML Docker container uses `https://pecl.php.net/get/yaml-2.2.0b2.tgz` instead of the Git version.

Comment (eldadfux) [reply Dockerfile:66]: Pecl was deprecated from the official Docker image: https://github.com/docker-library/php/issues/1029

Comment (SoftCreatR) [reply Dockerfile:66]: That's something else ;)

File: Dockerfile
Comment (SoftCreatR) [inline Dockerfile:213]: Missing newline ;)

File: Dockerfile
Comment (SoftCreatR) [inline Dockerfile:35]: This is not required to build Imagick (if that's the reason, why you install it).

File: Dockerfile
Comment (SoftCreatR) [inline Dockerfile:141]: Are `docker-cli` and `docker-compose` necessary within the container?

File: Dockerfile
Comment (eldadfux) [reply Dockerfile:35]: Are you talking about `imagemagick`, `imagemagick-dev` or both?

File: Dockerfile
Comment (eldadfux) [reply Dockerfile:141]: Yes. We're using them for both cloud function and the installation command. We might split them into separate dockerfiles in the future, but for now it's easier to maintain as one.

File: Dockerfile
Comment (SoftCreatR) [reply Dockerfile:35]: Just the dev package.

File: Dockerfile
Comment (eldadfux) [reply Dockerfile:66]: Yeh I know, just trying to keep the codebase consistent, I'll improve docs later.

**Review by SoftCreatR — APPROVED** [review]

**Review by TorstenDittmann — APPROVED** [review]
Some changes are a bit beyond my PHP horizon, but I can't see anything that catches my eye.

**Comment by eldadfux** [comment]
Thanks for all the help @SoftCreatR and @TorstenDittmann!

I'll wait for the official release before finally merging this PR.

**Review by SoftCreatR — APPROVED** [review]

**Review by christyjacob4 — APPROVED** [review]
Looks good @eldadfux

**Comment by TorstenDittmann** [comment]
- Updated to latest dependencies for
    - Composer
    - PHP Extensions
- Improved Dockerfile by cloning the desired branches right away without history for the PHP extensions

**Comment by eldadfux** [comment]
@TorstenDittmann ping me when ready for merge

**Comment by TorstenDittmann** [comment]
After https://github.com/appwrite/appwrite/pull/1101 is merged I will do some testing and tomorrow it should be ready 🙌

**Comment by eldadfux** [comment]
> After #1101 is merged I will do some testing and tomorrow it should be ready 🙌

Awesome, lets try to get the benchmark done tomorrow as well as it is still fresh...

File: .travis-ci/prepare.sh
Comment (eldadfux) [inline .travis-ci/prepare.sh:2]: Any reason why we can't have this just as part of the Travis YAML?

File: CHANGES.md
Comment (eldadfux) [inline CHANGES.md:20]:
```suggestion
- Upgraded to PVP 8.0 (#713)
+ Upgraded to PHP 8.0 (#713)
```

File: app/workers/functions.php
Comment (eldadfux) [inline app/workers/functions.php:341]: Should we create breaking change and make `_ENV_` => `_RUNTIME_`?

File: app/workers/webhooks.php
Comment (eldadfux) [inline app/workers/webhooks.php:42]: Leftovers?

File: docker-compose.yml
Comment (eldadfux) [inline docker-compose.yml:302]: _APP_FUNCTIONS_ENVS -> _APP_FUNCTIONS_RUNTIMES?

File: .travis-ci/prepare.sh
Comment (TorstenDittmann) [reply .travis-ci/prepare.sh:2]: Didn't wanna include this logic into the travis file for future adjustments we can add.

Both is fine with me 👍🏻

Comment (eldadfux) [reply .travis-ci/prepare.sh:2]: I think for now inside the Travis will be a bit less complex and not open the door to a lot more side-logic

File: app/config/variables.php
Comment (eldadfux) [inline app/config/variables.php:389]: Maybe we should add a new key, and mention the old one is deprecated on 0.8 (and push it to the bottom)

Comment (TorstenDittmann) [reply app/config/variables.php:389]: Added 👍🏻

**Pull request merged**
